### PR TITLE
Keyless AWS access support

### PR DESCRIPTION
**Pull request by nezihyigitbasi**
This PR adds a new parameter ``hive.s3.use-instance-credentials`` to enable
keyless AWS access using instance credentials.

### CODE REVIEW

**Comment by nezihyigitbasi** [comment]
@zhenxiao will cleanup & rebase #2375 after this PR gets merged.
@electrum This PR also addresses your comments in #2375

**Comment by colinmarc** [comment]
Huge +1 to this!

**Comment by nezihyigitbasi** [comment]
@electrum did you have a chance to look at this PR?

**Comment by nezihyigitbasi** [comment]
Rebased to master.

**Comment by mombergm** [comment]
@nezihyigitbasi Do you think we can add the ability to detect the availability of roles? Rather than setting whether to use them or not? More along the lines of #1938 ?

I completely support the ability to set a config param, but isn't it also a good idea to auto detect whether roles are available?

**Comment by colinmarc** [comment]
Here's one way to use the Hadoop credentials, but fallback to the default provider chain (which includes instance roles): https://gist.github.com/colinmarc/de4fc8e2f2805a3e02e9

**Comment by nezihyigitbasi** [comment]
@mombergm I would prefer to make it automatic, but this was @electrum 's comment to add this new flag, see #1948

**Comment by electrum** [comment]
@martint, do you have any opinion on if this should be enabled or disabled by default? I think it is dangerous to be enabled by default, but @dain and others disagree.

**Comment by mombergm** [comment]
I don't agree that it would be dangerous to enable by default. AWS EC2 Roles are specifically made for security reasons, so that your application can use Temporary Access Credentials provided by the VM and managed by the IAM service. The "Danger" here is the same as in any other service with the way you use it. If you grant access to the EC2 instance to the wrong people who can misuse those permissions its a people issue, not a software security issue.

**Comment by nezihyigitbasi** [comment]
@electrum why do you think it will be dangerous if enabled by default? I think it is more dangerous (in terms of security) to use secret keys by default than IAM roles.

**Comment by mandusm** [comment]
Technically, both static defined credentials and EC2 roles use credential pairs. The main difference is that the EC2 provided credentials are only temporary and gets rotated on a regular basis for security reasons. Which means that if your temporary keys are for whatever reason leaked, it will only work temporarily and cause you limited pain.

 As opposed to the static defined credentials in an config file somewhere which stays the same until revoked using the IAM service. If this gets leaked, you will need to revoke the credentials, but first you need to be aware that the credentials were leaked. At which point you might already be in bad pain.

The other level of security is the access to the instance itself. A person will not have access to the temporary credentials generated by the EC2 instance unless they have access to the instance itself. They need to be logged into the instance first. 

AWS Best Practices from all Services including their Security Advisory team is to always prefer the Temporary Credentials provided by the EC2 Service. Statically defining credentials into a config file is never a good idea. 

Therefore, needing to specify a config option to use temporary credentials goes against this methodology. I agree that the option should be there to disable using EC2 provided credentials, but the option to use credentials should by default be true, or preferred. The option should be to specifically disable it, not enable it.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestPrestoS3FileSystem.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    private AWSCredentialsProvider getAwsCredentialsProvider(PrestoS3FileSystem fs) throws NoSuchFieldException, IllegalAccessException
+    {
+        AmazonS3 s3 = fs.getS3Client();
+        return  (AWSCredentialsProvider) readPrivateField(s3, "awsCredentialsProvider");
+    }
+
+    private Object readPrivateField(Object obj, String fieldName) throws NoSuchFieldException, IllegalAccessException
+    {
+        Field f = obj.getClass().getDeclaredField(fieldName);
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestPrestoS3FileSystem.java]: We don't abbreviate variable names, so call this `field`.

**Comment by electrum** [comment]
I'm fixing this up and merging it. The comments are FYI.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestPrestoS3FileSystem.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import java.lang.reflect.Field;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import static org.testng.Assert.assertTrue;
+
+public class TestPrestoS3FileSystem
+{
+    @Test
+    public void testCreateAmazonS3Client() throws URISyntaxException, IOException, NoSuchFieldException, IllegalAccessException
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestPrestoS3FileSystem.java]: For tests we just say "throws Exception"

File: presto-hive/src/test/java/com/facebook/presto/hive/TestPrestoS3FileSystem.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        assertTrue(awsCredentialsProvider instanceof StaticCredentialsProvider);
+
+        conf = new Configuration();
+        conf.setBoolean(PrestoS3FileSystem.S3_USE_AWS_INSTANCE_CREDENTIALS, true);
+        fs.initialize(uri, conf);
+        awsCredentialsProvider = getAwsCredentialsProvider(fs);
+        assertTrue(awsCredentialsProvider instanceof InstanceProfileCredentialsProvider);
+    }
+
+    private AWSCredentialsProvider getAwsCredentialsProvider(PrestoS3FileSystem fs) throws NoSuchFieldException, IllegalAccessException
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestPrestoS3FileSystem.java]: Java 7 added parent class for these so you can say `throws ReflectiveOperationException`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestPrestoS3FileSystem.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    public void testCreateAmazonS3Client() throws URISyntaxException, IOException, NoSuchFieldException, IllegalAccessException
+    {
+        PrestoS3FileSystem fs = new PrestoS3FileSystem();
+        URI uri = new URI("s3n://test-bucket/");
+        Configuration conf = new Configuration();
+        conf.set("fs.s3n.awsSecretAccessKey", "test_secret_access_key");
+        conf.set("fs.s3n.awsAccessKeyId", "test_access_key_id");
+        fs.initialize(uri, conf);
+        AWSCredentialsProvider awsCredentialsProvider = getAwsCredentialsProvider(fs);
+        assertTrue(awsCredentialsProvider instanceof StaticCredentialsProvider);
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestPrestoS3FileSystem.java]: We have an `assertInstanceOf` method in the Airlift class `Assertions` (in the testing module).

**Comment by electrum** [comment]
Merged, thanks!

**Pull request closed without merging**